kimchi/src/folding/error.rs

```diff
@@ -0,0 +1,331 @@
+            })
+        }
+        Add(e1, e2) => {
+            let (a, b) = eval_bin(e1, e2);
+            EvalLeaf::bin_op(a, b, |a, b| *a + b, |a, b| *a += b)
+        }
+        Sub(e1, e2) => {
+            let (a, b) = eval_bin(e1, e2);
+            EvalLeaf::bin_op(a, b, |a, b| *a - b, |a, b| *a -= b)
+        }
+        Mul(e1, e2) => {
+            let (a, b) = eval_bin(e1, e2);
+            EvalLeaf::bin_op(a, b, |a, b| *a * b, |a, b| *a *= b)
+        }
+    }
+}
+pub(crate) fn eval_exp_error<'a, C: FoldingConfig>(
+    exp: &FoldingExp<C>,
+    env: &'a ExtendedEnv<C>,
+    side: Side,
+) -> EvalLeaf<'a, Fi<C>> {
+    let degree = exp.folding_degree();
+    use FoldingExp::*;
+
+    let eval_bin = |e1, e2| (eval_exp_error(e1, env, side), eval_exp_error(e2, env, side));
+    match exp {
+        Cell(col) => env.col(col, side),
+        Double(e) => {
+            let col = eval_exp_error(e, env, side);
+            col.map(Field::double, |f| {
+                Field::double_in_place(f);
+            })
+        }
+        Square(e) => match degree {
+            Degree::Two => {
+                let a = eval_exp_error(e, env, side);
+                let b = eval_exp_error(e, env, !side);
+                let cross = EvalLeaf::bin_op(a, b, |a, b| *a * b, |a, b| *a *= b);
+                cross.map(Field::square, |f| {
+                    Field::double_in_place(f);
+                })
+            }
+            _ => {
+                let e = eval_exp_error(e, env, side);
+                e.map(Field::square, |f| {
+                    Field::square_in_place(f);
+                })
+            }
+        },
+        Add(e1, e2) => {
+            let (a, b) = eval_bin(e1, e2);
+            EvalLeaf::bin_op(a, b, |a, b| *a + b, |a, b| *a += b)
+        }
+        Sub(e1, e2) => {
+            let (a, b) = eval_bin(e1, e2);
+            EvalLeaf::bin_op(a, b, |a, b| *a - b, |a, b| *a -= b)
+        }
+        Mul(e1, e2) => match (degree, e1.folding_degree()) {
+            (Degree::Two, Degree::One) => {
+                let a = eval_exp_error(e1, env, side);
+                let b = eval_exp_error(e2, env, !side);
+                let first = EvalLeaf::bin_op(a, b, |a, b| *a * b, |a, b| *a *= b);
+                let a = eval_exp_error(e1, env, !side);
+                let b = eval_exp_error(e2, env, side);
+                let second = EvalLeaf::bin_op(a, b, |a, b| *a * b, |a, b| *a *= b);
+                EvalLeaf::bin_op(first, second, |a, b| *a + b, |a, b| *a += b)
+            }
+            _ => {
+                let (a, b) = eval_bin(e1, e2);
+                EvalLeaf::bin_op(a, b, |a, b| *a * b, |a, b| *a *= b)
+            }
+        },
+    }
+}
+
+pub(crate) fn compute_error<C: FoldingConfig>(
+    exp: &IntegratedFoldingExpr<C>,
+    env: &ExtendedEnv<C>,
+    u: (Fi<C>, Fi<C>),
+) -> [Vec<Fi<C>>; 2] {
+    let add = |a, b| EvalLeaf::bin_op(a, b, |a, b| *a + b, |a, b| *a += b);
```

Comment:
It would be cleaner to implement `Add`, `Sub`, `Mul`, etc. for `EvalLeaf` and use those.